fv3core/stencils/remapping_part2.py

```diff
@@ -27,125 +27,176 @@ def init_phis(hs: FloatField, delz: FloatField, phis: FloatField, te_2d: FloatFi
-        # TODO pfull is a 1d var
-        kmp = grid.npz - 1
-        for k in range(pfull.shape[0]):
-            if pfull[k] > 10.0e2:
-                kmp = k
-                break
-    if last_step and not do_adiabatic_init:
-        if consv > constants.CONSV_MIN:
-            raise NotImplementedError(
-                "We do not support consv_te > 0.001 "
-                "because that would trigger an allReduce"
-            )
-        elif consv < -constants.CONSV_MIN:
-            raise Exception(
-                "Unimplemented/untested case consv("
-                + str(consv)
-                + ")  < -CONSV_MIN("
-                + str(-constants.CONSV_MIN)
-                + ")"
-            )
+class Remapping_Part2:
+    def __init__(self):
+        self.grid = spec.grid
+        self.namelist = spec.namelist
+        self._saturation_adjustment = utils.cached_stencil_class(SatAdjust3d)(
+            cache_key="satadjust3d"
+        )
+        self._copy_from_below_stencil = FrozenStencil(
+            copy_from_below,
+            origin=self.grid.compute_origin(),
+            domain=self.grid.domain_shape_compute(),
+        )
 
-    if spec.namelist.do_sat_adj:
-
-        kmp_origin = (grid.is_, grid.js, kmp)
-        kmp_domain = (grid.nic, grid.njc, grid.npz - kmp)
-        saturation_adjustment(
-            te,
-            qvapor,
-            qliquid,
-            qice,
-            qrain,
-            qsnow,
-            qgraupel,
-            qcld,
-            hs,
-            peln,
-            delp,
-            delz,
-            q_con,
-            pt,
-            pkz,
-            cappa,
-            r_vir,
-            mdt,
-            fast_mp_consv,
-            last_step,
-            akap,
-            kmp,
+        self._moist_cv_last_step_stencil = FrozenStencil(
+            moist_cv.moist_pt_last_step,
+            origin=(self.grid.is_, self.grid.js, 0),
+            domain=(self.grid.nic, self.grid.njc, self.grid.npz + 1),
         )
-        if fast_mp_consv:
-            sum_te(te, te0_2d, origin=kmp_origin, domain=kmp_domain)
-    if last_step:
-        moist_cv.compute_last_step(
-            pt, pkz, dtmp, r_vir, qvapor, qliquid, qice, qrain, qsnow, qgraupel, gz
+
+        self._basic_adjust_divide_stencil = FrozenStencil(
+            basic.adjust_divide_stencil,
+            origin=self.grid.compute_origin(),
+            domain=self.grid.domain_shape_compute(),
         )
-    else:
-        basic.adjust_divide_stencil(
-            pkz, pt, origin=grid.compute_origin(), domain=grid.domain_shape_compute()
+
+        self.do_sat_adjust = self.namelist.do_sat_adj
+
+        # TODO : When remapping becomes an object, these stencils can be declared in
+        #        __init__ when (I think) pFull can be passed into __init__
+        # self._saturation_adjustment_stencil = FrozenStencil(SatAdjust3d,...)
+        # self._sum_te_stencil = FrozenStencil(sum_te,...)
```

Comment:
I know I had said it would make the most sense to do this when remapping is an object, but actually you could do it now. where this object is initialized, -- you could pass pfull in as an argument, then compute kmp in this init method. e.g. remap_part2_compute = utils.cached_stencil_class(remap_part2.Remapping_Part2)(pfull, cache_key="remapping_part2") .
If you prefer to leave that to a future PR, that's fine with me.